lib/db/observed.go

```diff
@@ -0,0 +1,155 @@
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	return err
+}
+
+func (db *Lowlevel) RemovePendingDevice(device protocol.DeviceID) {
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	if err := db.Delete(key); err != nil {
+		l.Warnf("Failed to remove pending device entry: %v", err)
+	}
+}
+
+// PendingDevices drops any invalid entries from the database after a
+// warning log message, as a side-effect.  That's the only possible
+// "repair" measure and appropriate for the importance of pending
+// entries.  They will come back soon if still relevant.
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	for iter.Next() {
+		keyDev := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		deviceID, err := protocol.DeviceIDFromBytes(keyDev)
+		var bs []byte
+		var od ObservedDevice
+		if err != nil {
+			goto deleteKey
+		}
+		if bs, err = db.Get(iter.Key()); err != nil {
+			goto deleteKey
+		}
+		if err = od.Unmarshal(bs); err != nil {
+			goto deleteKey
+		}
+		res[deviceID] = od
+		continue
+	deleteKey:
+		l.Infof("Invalid pending device entry, deleting from database: %x", iter.Key())
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) AddOrUpdatePendingFolder(id, label string, device protocol.DeviceID) error {
+	key, err := db.keyer.GeneratePendingFolderKey(nil, device[:], []byte(id))
+	if err != nil {
+		return err
+	}
+	of := ObservedFolder{
+		Time:  time.Now().Round(time.Second),
+		Label: label,
+	}
+	bs, err := of.Marshal()
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	return err
+}
+
+// RemovePendingFolder removes entries for specific folder / device combinations, or all
+// combinations matching just the folder ID, when given an empty device ID.
+func (db *Lowlevel) RemovePendingFolder(id string, device []byte) {
+	if len(device) > 0 {
+		key, err := db.keyer.GeneratePendingFolderKey(nil, device, []byte(id))
+		if err != nil {
+			return
+		}
+		if err := db.Delete(key); err != nil {
+			l.Warnf("Failed to remove pending folder entry: %v", err)
+		}
+	} else {
+		iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+		if err != nil {
+			l.Warnf("Could not iterate through pending folder entries: %v", err)
+			return
+		}
+		defer iter.Release()
+		for iter.Next() {
+			if id != string(db.keyer.FolderFromPendingFolderKey(iter.Key())) {
+				continue
+			}
+			if err := db.Delete(iter.Key()); err != nil {
+				l.Warnf("Failed to remove pending folder entry: %v", err)
+			}
+		}
+	}
+}
+
+// PendingFolders drops any invalid entries from the database as a side-effect.
+func (db *Lowlevel) PendingFolders() (map[string]map[protocol.DeviceID]ObservedFolder, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
```

Comment:
The key is device first, right? Then you could pass the device as an argument and use it to create a prefix iterator selecting only the requested devices. Thus not needing to filter in model.

Reply:
I thought about that, but finally decided to keep that out of the DB layer.  This way we the result of `db.PendingFolders()` can in principle be cached in memory and reused for different filter criteria.

Reply:
You can keep that possibility by using the same semantics as on model: `EmptyDeviceID` means all devices.

Reply:
Yes sure.  Just felt cleaner to limit the complexity within `db` package at some point.  Do you really want it moved over, or is the current separation also acceptable?

Reply:
Yes. moved over please, because it's less complex given entries are stored in individual device->folder keys already.